### ISSUE TITLE
Completed event not fired on IOS

### ISSUE THREAD

**Issue by BenDevelopment** [xamarin/Xamarin.Auth#151]
The Completed event is not fired on IOS, Safari keep navigate to the callback url without telling what happened.
The token cannot be catched.
```
public override void ViewDidAppear(bool animated)
{
    base.ViewDidAppear(animated);

    if (_isShown) return;
    _isShown = true;

    var auth = AuthenticationHelper.CreateOAuth2Authenticator(Provider);
    auth.Completed += OnAuthenticationCompleted;
    var view = auth.GetUI() as UINavigationController;
    PresentViewController(view, true, null);
}

async void OnAuthenticationCompleted(object sender, AuthenticatorCompletedEventArgs e)
{
    // NOT FIRED HERE
    if (e.IsAuthenticated)
    {
        var access_token = e.Account.Properties.FirstOrDefault(i => i.Key.Equals("access_token")).Value ?? string.Empty;
        await AuthenticationHelper.SocialCallback(Provider, access_token);
    }
    else
        await Navigator.Instance.CloseAuthentication();
}
```

**Comment by moljac** [comment]
Can you provide details? nuget version and how OAuth2Authenticator is constructed?

thanks

**Comment by BenDevelopment** [comment]
OAuth2Authenticator is constructed with this method:

```
public static OAuth2Authenticator CreateOAuth2Authenticator(Provider provider)
{
    OAuthConfiguration configuration =
        provider == Provider.Google ? OAuthGoogle :
        provider == Provider.Facebook ? OAuthFacebook :
        null;

    return new OAuth2Authenticator(configuration.ClientId, configuration.Scope, new Uri(configuration.AuthorizeUrl), new Uri(configuration.RedirectUrl));
}

public enum Provider
{
    Facebook,
    Google,
    Twitter
}

public static OAuthConfiguration OAuthGoogle
{
    get
    {
        return new OAuthConfiguration()
        {
            ClientId = "clientid.apps.googleusercontent.com",
            Scope = "https://www.googleapis.com/auth/userinfo.email",
            AuthorizeUrl = "https://accounts.google.com/o/oauth2/auth",
            RedirectUrl = "http://staging.logwod.me/social-callback/google",
        };
    }
}


public static OAuthConfiguration OAuthFacebook
{
    get
    {
        return new OAuthConfiguration()
        {
            ClientId = "clientid",
            Scope = "email",
            AuthorizeUrl = "https://www.facebook.com/dialog/oauth",
            RedirectUrl = "http://staging.logwod.me/social-callback/facebook",
        };
    }
}

```
The nuget version is 1.3.2.5.

I've forgot to say that it works well on Android and UWP. The problem is only on IOS.

**Comment by moljac** [comment]
Will check tomorrow in the morning. This parts are going through some changes (WinRT is right now target and native UI - Chrome Custom Tabs and SFSafariViewController), so I might have messed something up.

So, UWP is working. Good news. Thanks.

**Comment by moljac** [comment]
@BenDevelopment 
This iOS issue: was it on simulator? Even more precise Simulator for iOS 10.x?

Storing to KeyChain (Add method ) returns status code -34018 on all iOS 10.x simulators.

Investigating proper [re]solution/workaround.

**Comment by BenDevelopment** [comment]
Yes this is on simulator 10.2.
But the error you are describing here is not the same error. The "-34018" occure when you try to save data with the account factory. On IOS you must have a Entitlements.plist file to use local storage. But this is not this error I'm facing. My problem is juste that the event is not fired. I have no error message.

**Comment by moljac** [comment]
In 10.2 simulator I had ceashes with Xmarin.Auth.Authentication error and code -34018.

Switching to simulator 9.2 and 8.4 I was getting UIAlerts.

And it seems it is not only Entitlements.plist which needs to be added. It seems that XCode needs at list one entry and thus people add Key Chain Sharing in capabilities

**Comment by BenDevelopment** [comment]
Yes you have to add Entitlements.plist and add it to Custom Entitlements in the project settings. But it only works if you have a Apple provisionning profile (it doesn't works with the free developper account), that's why I 'm using a different tool to save datas in local storage.
If you have alerts with auth.GetUI() (I don't know if this is that you are talking about when you say UIAlerts), you have to add some modifications in the renderer (the "get started" guide seems to be out of date).
Try to replace:
`PresentViewController (auth.GetUI (), true, null);`
By:
```
UIViewController authView = auth.GetUI() as UIViewController;
PresentViewController(authView, true, null);
```

**Comment by moljac** [comment]
@BenDevelopment 

I'm sorry, but I'm trying to answer bunch of questions and do some coding.

1. I'm using UIAlerts to show user token in Completed event handler.    
    This was what I was thinking with UIAlert
2. Component (and thus getting-started.md) is outdated. packaging component with Windows support and espacially samples is not trivial task anymore.

So this is new iOS API call:

        System.Object ui_controller_as_object = auth.GetUI();
        if (auth.IsUsingNativeUI == true)
        {
            SafariServices.SFSafariViewController c = null;
            c = (SafariServices.SFSafariViewController)ui_controller_as_object;
            PresentViewController(c, true, null);
        }
        else
        {
            UIViewController c = (UIViewController)ui_controller_as_object;
            PresentViewController(c, true, null);
        }

Yes it is breaking change System.Object is returned!!

And here I subscribe to event

https://github.com/xamarin/Xamarin.Auth/blob/portable-bait-and-switch/samples/Traditional.Standard/references01projects/Providers/Xamarin.Auth.Sample.XamarinIOS/TestProvidersController.cs#L91

And handle event (here be UIAlerts)

https://github.com/xamarin/Xamarin.Auth/blob/portable-bait-and-switch/samples/Traditional.Standard/references01projects/Providers/Xamarin.Auth.Sample.XamarinIOS/TestProvidersController.cs#L166-L244

See this issue with more details for NEW API.

https://github.com/xamarin/Xamarin.Auth/issues/146#issuecomment-284730760

**Comment by mkroghj** [comment]
I can confirm that it seems like the completed eventhandler is NOT called when the browser is redirected to redirectUrl, using version 1.3.2.5 and iOS 10.2 simulator.

I have the following code:

`						auth.AllowCancel = true;
						auth.IsUsingNativeUI = true;
						auth.Completed += completedHandler;
						var ui = auth.GetUI();
						PresentViewController ((SafariServices.SFSafariViewController)ui, true, null);
`

but the completedHandler is not called. The view just shows the redirectUrl, but no event is fired. There are no errors reported, since I do not save anything to the keychain.

**Comment by DFerenczi** [comment]
Same for me. Completed-event does not fire on iOS, tried Simulator iPhone 6 iOS 9.1, Iphone 5s iOS 9.1, iPhone 7 iOS 10.2 and in none of them is the event fired. 

Tried with both: 
PresentViewController((SafariServices.SFSafariViewController)oauth.GetUI(), true, null);

and

PresentViewController((UIViewController)oauth.GetUI(), true, null);

Thanks!

**Comment by BenDevelopment** [comment]
Any idea  of how to fix this ?

**Comment by moljac** [comment]
@BenDevelopment @DFerenczi @mkroghj 

So, my tests iOS != 10.x, nuget 1.3.2.5

(abandoned search for 10.x issues until new features are implemented)

* FaceBook WWW type app Redirect URL http[s]://xamarin.com        
   OK
* FaceBook WWW type app Redirect URL http[s]://localhost        
   OK (getting access_token with error trying to load localhost!)
* FaceBook WWW type app Redirect URL http[s]://127.0.0.1        
   OK (getting access_token with error trying to load localhost!)
* github app Redirect URL http://xamarin.com        
   OK
* github app Redirect URL https://xamarin.com        
   OK
* github app Redirect URL https://localhost        
   OK
* github app Redirect URL https://127.0.0.1        
   OK

NOTE: 

* support for hosts like localhost and 127.0.0.1 is NEW
* NOT using custom url scheme[s]! This was never supported in Xamarin.Auth and this is under heavy development right now. Thus SFSafariViewController will NOT work (it might if you are lucky, but I would not bet).

**Comment by mkroghj** [comment]
@moljac Sorry, I may be misunderstanding - so you are not testing on 10.x until new features are added? I think Apple announced in January, that over 75 % of all of their active devices are now using iOS 10. Seems to me that this should be prioritised.

**Comment by rasweh** [comment]
@moljac We have a finished iOS app, ready to be put into App Store - except for users not being able to authenticate because of the bug reported by BenDevelopment.
I agree with mkroghj: as just about 80% (as of February 20th this year, see https://developer.apple.com/support/app-store/) of iOS users now run iOS 10.x this is a major bug.
Hopefully you'll be able to squeeze out the bug very soon. Have you got a time estimate?

**Comment by Jake-Barrow** [comment]
The functionality of Xamarin.Auth 1.3.2.5 is not fully implemented. The best course of action is to downgrade to 1.3.2.1 (the last stable before new partial functionality made it into the package)

The changes being made refer to the Google Authentication changes happening on 20th April 2017 (https://developers.googleblog.com/2016/08/modernizing-oauth-interactions-in-native-apps.html)

Hopefully the new features will be released by then, but if you can't wait, downgrade to 1.3.2.1 and use the old authentication method. Other than that there is nothing you can do until the functionality is finished.

**Comment by moljac** [comment]
@Jake-Barrow 

Users can not use this functionality by default (implicitly) and especially should not for production. 
This features are not turned on (yet), but they will be.

User needs to explicitly turn on Native UI in constructor to obtain SFSafariViewController and Custom Tabs.

New packages have this for testing, but it is undocumented and unstable. Old docs for Xamarin.Auth component are still valid except the cast and nuget dependency on Android.Support.Customtabs and this dependency will be removed (Android only).

And now how did you deduce that this bug is about new feature?

**Comment by moljac** [comment]
Fixed in 1.3.2.6.

My mistake and I sincerely apologize. All this hectic with new Windows support and NativeUI made me push wrong logic (new NativeUI support turned on in the library). Those users that were using explicit constructor parameter to turn off NativeUI did not have issues.

Please let me know the results.

Now some details:

@mkroghj 

> @moljac Sorry, I may be misunderstanding - 

Yes. You misunderstood: those tests did not cover iOS 10.x simulator on that day (Thursday last week). Explanation why I do not use iOS 10.x on daily basis is below.

> so you are not testing on 10.x until new features are added? 

I'm testing 10.x on device and to keep order in my life - iOS device tests beginning of the week. Why because I'm covering iOS, Android and several Windows Platforms right now and if I would go and redo all test on all platforms both simulators/emulators and devices I wouldn't write single line of new code. 

The reason for not testing iOS 10.x on simulator is simply I'm experiencing (in most, but not all cases) issues described here:
 
*   139 Could not save account to KeyChain: -34018 
     https://github.com/xamarin/Xamarin.Auth/issues/139
*   128 Error: Could not save account to KeyChain -- iOS 10
     https://github.com/xamarin/Xamarin.Auth/issues/128

Issue comment which shows that we have to help users with really basic things:

https://github.com/xamarin/Xamarin.Auth/issues/128#issuecomment-285351561

So, 2 issues for the same bug and bug which has NOTHING to do with Xamarin. To prove that few links from Apple developer forums, stackoverflow and some products hit by this bug (hockeyapp):

https://forums.developer.apple.com/message/179846
https://forums.developer.apple.com/thread/60617
https://forums.developer.apple.com/thread/51071
https://forums.developer.apple.com/thread/62847
http://stackoverflow.com/questions/38456471/secitemadd-always-returns-error-34018-in-xcode-8-in-ios-10-simulator
https://support.hockeyapp.net/kb/client-integration-ios-mac-os-x-tvos/known-issues-with-ios#hockeysdk-ios-and-ios-10-simulators
https://support.hockeyapp.net/discussions/problems/63710-the-authentication-token-could-not-be-stored-due-to-a-keychain-error

So, I have added exception that in Message explains user to add Entitlements.plist, but this is semi-workaround, because adding Entitlements just for simulator causes problems for device. (Read apple devs' opinions). So, I still have no solution for iOS 10.x simulator and I have XCode 8.2, so the problem is not fixed yet.

> I think Apple announced in January, that over 75 % of all of their active devices are now using 
> iOS 10. 

OK. But quoting numbers does not increase my opinion on the iOS 10 importance. I'm aware of it. 

> Seems to me that this should be prioritised.

So, today is Monday and I just tested 1.3.2.5 on 2 iOS 10.x devices and I was able to get token from Facebook even for http[s]://127.0.0.1 and http[s]://localhost redirect_urls. Yes I get OnError first that UIWebView cannot connect to server (of course), but OnError event is triggered and after that onComplete.

In 1.3.2.6 which is  published, this is fixed so no error is raised. I was using explicit constructor parameter and people that read my samples could deduce and turn old API on (and new off).
My mistake.

@rasweh 

> @moljac We have a finished iOS app, ready to be put into App Store - except for users not being
> able to authenticate 

OK. 1st I'm sorry that you are/were blocked. Please test 1.3.2.6 and report if this fixes the issue.

> Hopefully you'll be able to squeeze out the bug very soon. 

Nuget 1.3.2.6 out since yesterday. Please report outcome/results.

> Have you got a time estimate?

For new API (NativeUI support)? 

I wish. I was afraid that Windows (WinRT both Windows Store and Windows Phone, UWP, Windows Phone 8.x Silverlight) support will bring influx of people to Xamarin.Auth and thus more issues and feature requests. And this happened and now I'm a bit under pressure on all fronts.

Generally my priority list:

1.  NativeUI support ASAP alpha of usable API for testing
2.  fix few issues in Android (most of them not critical)
3.  component packaging (including docs)
4.  Windows support (few nice-to-haves)
5.  new features (some ideas like fingerprint etc)

**Comment by roopeshnaik** [comment]
I am using the 1.3.2.7 and the auth completed is not fired for Android,it just redirects to url and doesnot start the activity

**Comment by moljac** [comment]
@sopoco @roopeshnaik 

I'm using 1.3.2.7 and OAuth2 events for 

*    Facebook Android and iOS
*    Microsoft Live Android and iOS   
      and 
*    Github (Explicit) iOS only (Android fires OnError)

are firing.

Twitter OAuth1 events are not firing and I'm working on that.

Current work in progress:

*   fixing Android code for OAuth1 and loops in OnError
*   cleaning Android code and adding Xamarin.Auth.Exceptions for errors with messages
*   native support

**Comment by roopeshnaik** [comment]
how can i handle the redirect uri from the custom chrome tabs

            auth.Completed += Auth_Completed;

            var ui_intent = auth.GetUI(this);
            if (auth.IsUsingNativeUI == true)
            {
                // using new Native UI API Chrome Custom Tabs on Android
                Uri uri = auth.GetInitialUrlAsync().Result;
                Android.Net.Uri uri_android = Android.Net.Uri.Parse(uri.AbsoluteUri);                

                CustomTabsActivityManager customTabs = new CustomTabsActivityManager(this);
                CustomTabsIntent cti = null;
                cti = (CustomTabsIntent)ui_intent;
                cti.LaunchUrl(this, uri_android);
            }

After authenticating it redirects to uri and do i need to Startactivity after cti.LaunchUrl(this, uri_android)

**Comment by Jake-Barrow** [comment]
@roopeshnaik The NativeUI support is not yet done, so no events are fired. For now you need to use the old UI (IsUsingNativeUI = false in ctor)

**Comment by roopeshnaik** [comment]
@Jake-Barrow , the Old Webviews are supposed to be deprecated/blocked by 20-04-2017, so this method will not work for Google Oauth2
https://developers.googleblog.com/2016/08/modernizing-oauth-interactions-in-native-apps.html

**Comment by moljac** [comment]
@roopeshnaik 

Could you be so kind and read threads/issues a bit before posting? Please.

I am (we are) aware of those facts and are working on solution. 

See my weekend work repo:

https://github.com/moljac/HolisticWare.Android.Support.CustomTabs.Chromium.SharedUtilities

This is being silently integrated added, so users can use Native OAuth + apply style and themes.

Instead of just stating "X is not working" provide a little bit more info.

**Comment by DFerenczi** [comment]
Nice to see that the completed callback is working again! GJ 👍 I don't know if this problem is somehow related to this, please tell me if I should create a new issue, but im always getting a popup saying "Error authenticating: invalid_request" even though the fetch is successful i.e. I get a valid access token. 

Any ideas? 

/Ferenczi

**Comment by moljac** [comment]
@DFerenczi 
It was working all the time, but users have turned on new API which is not completely finished. I tried to prepare more flexible custom schemes support, but this will not be possible (http and https in Custom Tabs and SFSafariViewController). Good thing is that new API is close to be finished.

Regarding your problem: I need more info. I presume this is Android. Am I right? I need to clean up some code it seems that it runs into error loop, but not 100% sure.
Which version of nuget? Which provider and OAuth protocol?

**Comment by DFerenczi** [comment]
Sound good! 

I get the error msg on both platforms (ios and android), i'm using 1.3.2.7 on iOS and 1.3.2.6 on Android while using OAuth2Authenticator.. I've tried to set the ShowErrors property to false just to avoid the popup but it still shows. Hope this helps in any way.

**Comment by ads90** [comment]
the event doesnt fired in the new version please fix it

**Comment by moljac** [comment]
@ads90 
1. which is new version?
2. can we see your ctor? (Are you using nativeUI)

**Comment by githubdigo** [comment]
iam using xamarin auth 1.4.1,the completed event is not getting fired after the navigation to redirection url.

**Comment by githubdigo** [comment]
hi please see the details of constructor  and safari navigation

auth2Request = new OAuth2Authenticator(
					clientId: clientIdGoogle, // Oauth2Request client id
					scope: scopeGoogle, // scopes
					authorizeUrl: new Uri(authorizeUrlGoogle), // the authorization URL f 
						isUsingNativeUI: true,
						redirectUrl: new Uri(redirectUrlGoogle)); // the redirect URL for 

nd safari navigation 


SafariServices.SFSafariViewController c = null;
						c = (SafariServices.SFSafariViewController)ui_controller_as_object;

						if (UIDevice.CurrentDevice.CheckSystemVersion(10, 0))
						{
							c.PreferredBarTintColor = UIColor.Gray;
							c.PreferredControlTintColor = UIColor.White;
						}
						else
						{
							c.View.TintColor = UIColor.Gray;
						}

						Action view_controller_customization =
							() =>
							{
								//c.NavigationController.NavigationBar.TintColor = color_xamarin_blue;
							};
						//------------------------------------------------------------
						// [REQUIRED] launching SFSafariViewController
						PresentViewController(c, true, view_controller_customization);
						//=================================================================

**Comment by JichinThomas** [comment]
@moljac I too face the issue that @githubdigo is faced with. Sorry to ask, but is there something (possibly trivial) that we are missing in common?

**Comment by moljac** [comment]
@JichinThomas @githubdigo 

Can you please provide:

*   AppDelegate.OpenUrl code
*   Info.plist and 
*   redirect_url

**Comment by JichinThomas** [comment]
_AppDelegate OpenUrl code_

Public override bool OpenUrl ( UIApplication application,NSUrl url,string sourceApplication, NSObject annotation){

    System.Text.StringBuilder sb = new System.Text.StringBuilder();
    sb.AppendLine("OpenURL Called");
    sb.Append("     url         = ").AppendLine(url.AbsoluteUrl.ToString());
    sb.Append("     application = ").AppendLine(sourceApplication); 
    sb.Append("     annotation  = ").AppendLine(annotation?.ToString());
    System.Diagnostics.Debug.WriteLine(sb.ToString()); 

    System.Uri uri = new Uri(url.AbsoluteString);
    IDictionary<string, string> fragment = Utilities.WebEx.FormDecode(uri.Fragment);
    Account account = new Account(
                            "username",
                             new Dictionary<string, string>(fragment) );
    return true; 

} 

_Info.plist_
CFBundleIdentifier     _com.xyz.candidate.qa_
CFBundleURLName     _com.xyz.candidate.qa_
CFBundleURLSchemes
                       _xamarin-auth
                       xamarinauth
                       xamarin.auth_
CFBundleURLTypes     _Viewer_


_Redirect URL used is http://xamarin.com_

**Comment by moljac** [comment]
@JichinThomas 

redirect_url!!! NO GOOD! 

iOS will intercept http redirect uri in Browser, so no events will be fired.

It must be: 
```
xamarin-auth
```
or 
```
xamarinauth
```
or
```
xamarinauth
```

And those are Xamarin.Auth schemes - you need to come up with your own!

And one more question: Which OAuth provider?

**Comment by JichinThomas** [comment]
Also, am getting an alert when trying to authenticate using Xamarin.Auth2.  Have you seen a similar one when trying Xamarin.Auth2 Native UI case ?
![Uploading warning.png…]()

**Comment by JichinThomas** [comment]
@moljac  Am using Google Sign in option using Xamarin.Auth2 1.4.1  version

**Comment by moljac** [comment]
cannot see image?

Is it something like:

                if (scheme.StartsWith("http", StringComparison.InvariantCultureIgnoreCase))
                {
                    StringBuilder sb = new StringBuilder();
                    sb.AppendLine("WARNING");
                    sb.AppendLine($"Scheme = {scheme}");
                    sb.AppendLine($"Native UI used with http[s] schema!");
                    sb.AppendLine($"Redirect URL will be loaded in Native UI!");
                    sb.AppendLine($"OAuth Data parsing might fail!");

                    ShowErrorForNativeUI(sb.ToString());
                }

As you can see I did put this WARNING for users to WARN them that they are using http or https schema with SFSafariViewController and that Redirect URL will NOT be loaded and OAuth data will NOT be parsed!!!!!

For google sign in this should be YOUR (AND ONLY YOURS) redirect_url

com.xyz.candidate.qa:/oauth2redirect

**Comment by JichinThomas** [comment]
@moljac 
yes, this is the warning.
Is the second part in the redirect url, the callback function ?

**Comment by moljac** [comment]
@JichinThomas Sorry, but I do not understand "second part"!

If you need faster support go to slack Xamarin Community chat in #xamarin-auth-social channel and ping us there. There are several people that have implemented new OAuth with google and other providers.

Using github issues as chat is at least overkill.

**Comment by biapar** [comment]
+1

**Comment by moljac** [comment]
@biapar ??

**Comment by biapar** [comment]
@moljac: the events are not intercepted with native flag on ( see Google reauirements ). I use xamarin firm and todoawsauth. Facebook auth is ok. Xam.auth lib 1.4.1

**Comment by moljac** [comment]
@biapar 
I bet your redirect-url is http://* or https://*.
You need custom scheme (registered), so your app intercepts opening page with custom scheme and parses url and raises events.

+1 does not help us at all. You need to provide more info if you want to be helped out, otherwise I must resort to ignore mode.

**Comment by biapar** [comment]
I redirect to my site http://. How to do the "custom scheme"?

**Comment by moljac** [comment]
@biapar 

> I redirect to my site http://

Another bet won!
This is exactly your problem - NativeUI CANNOT work with http[s] schemes!
OK. It might/could, but with a lot of effort and no guarantees for success.

> How to do the "custom scheme"?

Sorry too long for answer. 

Check:
https://github.com/moljac/Xamarin.Auth.Samples.NugetReferences/blob/master/Xamarin.Forms/Evolve16Labs/Portable/MainPage.xaml.cs#L406
and
https://github.com/moljac/Xamarin.Auth.Samples.NugetReferences/blob/master/Xamarin.Forms/Evolve16Labs/Portable/MainPage.xaml.cs#L412

Custom scheme is anything that is not here:
https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml

**Comment by biapar** [comment]
Thank you. Why Do you use different redirect URI for IOS and Android?

**Comment by biapar** [comment]
What type of client OAUTH Do you created into Google Console? I created web application.

**Comment by moljac** [comment]
@biapar 

> Thank you. Why Do you use different redirect URI for IOS and Android?

Good. Very good question.
Do you remember creating Credentials for Android and iOS app?
You had to submit Package ID on Android (+ SHA) and Bundle Id on iOS. OK. Some developers could make those match, but most of the IDEs (XCode, Android Studio, Xamarin Studio, Visual Studio) will make it based on project settings - project name user defines during project creation.
Based on Project Id and/or Bundle Id google creates schema for callback url. Why? My speculation is to track your app[s].

> What type of client OAUTH Do you created into Google Console?

Android, iOS Other.
Those are (according to google) Installed apps.

>I created web application.

No. Wrong. Web App is let say for server (ASP.net app that has login with google. Saving secure data there is considered OK (debatable), but on mobile it is not.
Xamarin.Auth used to work with Web Apps only (needed real redirect-url, not localhost,or custom schemes), but this has changed

**Comment by moljac** [comment]
Closing.

**Comment by Bmahecha123** [comment]
So for Google Sign In using this instructional https://developer.xamarin.com/samples/xamarin-forms/WebServices/OAuthNativeFlow/ takes me to google sign in, once I sign in my browser stays open and the OnAuthCompleted event never get's fired. I set the breakpoint and it never gets there. So things seem to be working to an extent.. Just need this OnAuthCompleted event to fire off.

`		void OnLoginClicked(object sender, EventArgs e)
		{
			string clientId = null;
			string redirectUri = null;

			switch (Device.RuntimePlatform)
			{
				case Device.iOS:
					clientId = Constants.iOSClientId;
					redirectUri = Constants.iOSRedirectUrl;
					break;

				case Device.Android:
					clientId = Constants.AndroidClientId;
					redirectUri = Constants.AndroidRedirectUrl;
					break;
			}

			var authenticator = new OAuth2Authenticator(
				clientId,
				null,
				Constants.Scope,
				new Uri(Constants.AuthorizeUrl),
				new Uri(redirectUri),
				new Uri(Constants.AccessTokenUrl),
				null,
				true);

			authenticator.Completed += OnAuthCompleted;
			authenticator.Error += OnAuthError;

			AuthenticationState.Authenticator = authenticator;

			var presenter = new Xamarin.Auth.Presenters.OAuthLoginPresenter();
			presenter.Login(authenticator);
		}

		async void OnAuthCompleted(object sender, AuthenticatorCompletedEventArgs e)
		{
			var authenticator = sender as OAuth2Authenticator;
			if (authenticator != null)
			{
				authenticator.Completed -= OnAuthCompleted;
				authenticator.Error -= OnAuthError;
			}

			User user = null;
			if (e.IsAuthenticated)
			{
				// If the user is authenticated, request their basic user data from Google
				// UserInfoUrl = https://www.googleapis.com/oauth2/v2/userinfo
				var request = new OAuth2Request("GET", new Uri(Constants.UserInfoUrl), null, e.Account);
				var response = await request.GetResponseAsync();
				if (response != null)
				{
					// Deserialize the data and store it in the account store
					// The users email address will be used to identify data in SimpleDB
					string userJson = await response.GetResponseTextAsync();
					user = JsonConvert.DeserializeObject<User>(userJson);
				}

				if (account != null)
				{
					store.Delete(account, Constants.AppName);
				}

				await store.SaveAsync(account = e.Account, Constants.AppName);
				await DisplayAlert("Email address", user.Email, "OK");
			}
		}

		void OnAuthError(object sender, AuthenticatorErrorEventArgs e)
		{
			var authenticator = sender as OAuth2Authenticator;
			if (authenticator != null)
			{
				authenticator.Completed -= OnAuthCompleted;
				authenticator.Error -= OnAuthError;
			}

			Debug.WriteLine("Authentication error: " + e.Message);
		}`

**Comment by judsonabraham301** [comment]
I'm facing the same issue that @Bmahecha123 is facing.

**Comment by mr5z** [comment]
For those who are facing an issue not being redirect to their app after the authentication, the docs is actually missing an essential piece of code. Here's the link: https://github.com/xamarin/Xamarin.Auth/blob/master/samples/Xamarin.Forms/Evolve16Labs/05-OAuth/iOS/Xamarin.Auth/AppDelegate.OpenUrl.cs

The `AuthenticationState` is an instance of OAuth2Authenticator.

**Comment by sschmidTU** [comment]
I found a solution that hasn't been mentioned:
Setting the HttpClient implementation to NsUrlSession instead of Managed.
(right click your project in the solution -> Options -> iOS build.

Managed is slower, doesn't support TLS 1.2 (needed for ATS, set as default on iOS9+).
The only downside to NSUrlSession is that it requires iOS7+. There's another client that requires iOS6+.

Before this, auth.Completed mostly fired right away but on every 2nd or 3rd attempt it took over a minute, while the user was stuck in the browser, logged into google, on the google start page.

More Details here:
https://stackoverflow.com/a/57875600/10295942

**Comment by mr5z** [comment]
> I found a solution that hasn't been mentioned:
> Setting the HttpClient implementation to NsUrlSession instead of Managed.
> (right click your project in the solution -> Options -> iOS build.
> 
> Managed is slower, doesn't support TLS 1.2 (needed for ATS, set as default on iOS9+).
> The only downside to NSUrlSession is that it requires iOS7+. There's another client that requires iOS6+.
> 
> Before this, auth.Completed mostly fired right away but on every 2nd or 3rd attempt it took over a minute, while the user was stuck in the browser, logged into google, on the google start page.
> 
> More Details here:
> https://stackoverflow.com/a/57875600/10295942

A much better solution for this is to completely ditch out this outdated library and switch to [Clancey Simple Auth](https://github.com/Clancey/SimpleAuth).
Took me days to realize this library wasn't worth the effort.

**Comment by Tushar5292** [comment]
Hello Everyone , Anybody can help me out that I am getting attached screen shot warning on last line of below code (presentor) I am testing this on simulator i.e on IOS. I have sent all the required parameters.  

![Simulator Screen Shot - iPhone 8 - 2020-09-30 at 13 29 25](https://user-images.githubusercontent.com/40889976/94660819-d65ba100-0323-11eb-9a79-c62f9c563321.png)


var authenticator = new OAuth2Authenticator(
                    clientId: clientId,
                    clientSecret: string.Empty,
                    scope: Constants.FacebookScope,
                    authorizeUrl: new Uri(Constants.FacebookAuthorizeUrl),
                    redirectUrl: new Uri(redirectUri),
                    accessTokenUrl: new Uri(Constants.FacebookAccessTokenUrl),
                    getUsernameAsync: null,
                    isUsingNativeUI: true
                    );
                    authenticator.Completed += OnAuthCompleted;
                    authenticator.Error += OnAuthError;

                    //authenticator.IsLoadableRedirectUri = true;

                    AuthenticationState.Authenticator = authenticator;

                    var presenter = new Xamarin.Auth.Presenters.OAuthLoginPresenter();
                    **presenter.Login(authenticator);**
